generative_models/cGAN_PyTorch/main.py

```diff
@@ -0,0 +1,177 @@
+    gen_imgs = generator(z, labels)
+    save_image(gen_imgs.data, args.outdir +"Image/%d.png" % batches_done, nrow=args.num_class, normalize=True)
+
+
+def train(args, generator, discriminator, dataloader, loss, img_type,label_type,gen_optimizer,d_optimizer,fix_label,fix_noise):
+    
+    generator.train()
+    discriminator.train()
+
+    G_losses = []
+    D_losses = []
+ 
+    for epoch in range(1, args.epochs + 1):
+        G_loss = 0.
+        D_loss = 0.
+        start_time = time.time()
+        for i, data in enumerate(dataloader):
+            (imgs, labels) = data
+                        
+            batch_size = imgs.shape[0]
+            #print(batch_size)
+            imgs = Variable(imgs.type(img_type)).to(args.device)
+            labels = Variable(labels.type(label_type)).to(args.device)
+
+            # Creating real and fake label for calculation of loss
+            r_label = Variable(img_type(batch_size, 1).fill_(0.9)).to(args.device)
+            f_label = Variable(img_type(batch_size, 1).fill_(0.0)).to(args.device)
+
+            # Training Generator
+
+            gen_optimizer.zero_grad()
+
+            noise = Variable(img_type(np.random.normal(0, 1, (batch_size, args.latent_dim)))).to(args.device)
+            rand_label = Variable(label_type(np.random.randint(0, args.num_class, batch_size))).to(args.device)
+            dis = discriminator(generator(noise, rand_label), rand_label)
+            # print(type(dis),'  ',type(r_label))
+            g_loss = loss(dis, r_label)
+            g_loss.backward()
+            gen_optimizer.step()
+
+            # Training Discriminator
+
+            d_optimizer.zero_grad()
+
+            noise = Variable(img_type(np.random.normal(0, 1, (batch_size, args.latent_dim)))).to(args.device)
+            rand_label = Variable(label_type(np.random.randint(0, args.num_class, batch_size))).to(args.device)
+
+            d_real = discriminator(imgs, labels)
+            loss_real = loss(d_real, r_label)
+
+            d_fake = discriminator(generator(noise, rand_label).detach(), rand_label)
+            #print(d_fake.shape," ",f_label.shape)
+            loss_fake = loss(d_fake, f_label)
+
+            d_loss = 0.5 * (loss_fake + loss_real)
+
+            d_loss.backward()
+            d_optimizer.step()
+
+            G_loss += g_loss.item()
+            D_loss += d_loss.item()
+
+           
+        print('Epoch {} || G_loss: {} || D_loss: {} || Time elapsed: {}'.format(epoch, G_loss / (i), D_loss / (i),time.time()-start_time))
+        G_losses.append(G_loss / (i))
+        D_losses.append(D_loss / (i))
+
+        sample_image(args,fix_noise,fix_label, epoch, generator)
+
+        # Checkpoint
+        torch.save(generator.state_dict(), args.outdir + 'generator_state/generator_{}_.pth'.format(epoch))
+        torch.save(discriminator.state_dict(),args.outdir + 'discriminator_state/discriminator_{}_.pth'.format(epoch))
+
+    return G_losses, D_losses
+
+def main():
```

Comment:
maybe position main() function above train() function, it is nice to have parser at the top for reference.